bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/LGWebOSHandlerFactory.java

```diff
@@ -53,7 +55,8 @@ public LGWebOSHandlerFactory(final @Reference WebSocketFactory webSocketFactory,
          * Cannot use openHAB's shared web socket client (webSocketFactory.getCommonWebSocketClient()) as we have to
          * change client settings.
          */
-        this.webSocketClient = webSocketFactory.createWebSocketClient("lgwebos");
+        var httpClient = new HttpClient(new SslContextFactory.Client(true));
```

Comment:
Question: would it be better to use HttpClientFactory ?

Reply:
I don't think it's possible, since there is no overload of `createHttpClient` taking an `SslContextFactory`. Also I can't see any set methods allowing to set it after the object has been created.

It seems to be a common approach:
```
$ grep -R "new HttpClient(new SslContextFactory"
bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java:        HttpClient httpClient = new HttpClient(new SslContextFactory.Client());
bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinHttpClientFactoryImpl.java:            httpClient = new HttpClient(new SslContextFactory.Client(true));
bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpHandlerFactory.java:        this.secureClient = new HttpClient(new SslContextFactory.Client());
bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpHandlerFactory.java:        this.insecureClient = new HttpClient(new SslContextFactory.Client(true));
bundles/org.openhab.binding.miio/src/test/java/org/openhab/binding/miio/internal/MiotJsonFileCreator.java:                httpClient = new HttpClient(new SslContextFactory.Client());
bundles/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/wwn/handler/WWNRedirectUrlSupplier.java:        HttpClient httpClient = new HttpClient(new SslContextFactory.Client());
bundles/org.openhab.binding.seneye/src/main/java/org/openhab/binding/seneye/internal/SeneyeService.java:    private HttpClient httpClient = new HttpClient(new SslContextFactory.Client());
bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java:            this.httpClient = new HttpClient(new SslContextFactory.Client(true));
bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoControlHandlerFactory.java:        httpClient = new HttpClient(new SslContextFactory.Client());
bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/Tr064HandlerFactory.java:        this.httpClient = new HttpClient(new SslContextFactory.Client(true));
bundles/org.openhab.binding.unifi/src/main/java/org/openhab/binding/unifi/internal/UniFiThingHandlerFactory.java:        httpClient = new HttpClient(new SslContextFactory.Client(true));
bundles/org.openhab.binding.venstarthermostat/src/main/java/org/openhab/binding/venstarthermostat/internal/handler/VenstarThermostatHandler.java:        httpClient = new HttpClient(new SslContextFactory.Client(true));
bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoHandlerFactory.java:        this.secureClient = new HttpClient(new SslContextFactory.Client(false));
bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoHandlerFactory.java:        this.insecureClient = new HttpClient(new SslContextFactory.Client(true));
```

Reply:
I believe I did here:
https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.remoteopenhab/src/main/java/org/openhab/binding/remoteopenhab/internal/RemoteopenhabHandlerFactory.java#L146

That being said, I don't know what is the best way, using or not httpClientFactory.

Reply:
Hmm, okay, proven possible. 👍 It seems your approach gives you more control by managing your own trust manager. In the case at hand I don't think we need that, so it seems overly complicated for the purpose? In both cases we end up having our own `HttpClient` instance.

Reply:
I show you the example not for the trust manager stuff but rather to show that it is possible and simple to create a HTTP client using httpClientFactory and then to associate to it a SSL context.

Reply:
But you will need to configure that SSL context in order to make it trust self-signed certificates like the parameter here (true = trustAll):

```java
new SslContextFactory.Client(true)
```

I might be missing something, and it's not fully within my comfort zone, so let me ask in another way: What is the advantage of the alternative approach of getting the instance from the factory and configuring the SSL context manually?

Reply:
No my confort zone too.

I just have a quick look at the `httpClientFactory` in the OH core framework. I can see that it creates a thread pool with a specific OH name (this name will reference your binding) and specific min/max values (OH parameters) for threads. There is also stuff regarding proxying but I am not confortable with that part.
So the interest of using this class is to have a HTTP client relying on a thread pool created by openHAB with specific name & custom settings.
At the reverse, I assume your option will create a thread pool with default settings customized by Jetty.

In case you are interested, you could probably try that:
```
var httpClient = httpClientFactory.createHttpClient(LGWebOSBindingConstants.BINDING_ID);
httpClient.getSslContextFactory().setSslContext(new SslContextFactory.Client(true));
```

Reply:
With httpClientFactory, the created HTTP client has its own thread pool.
Without httpClientFactory, I don't know if a dedicated thread pool is created or if a global Jetty thread pool is used.

Reply:
By the way, your code is also correct. I am just curious and would be interested to know what is the recommended option for OH bindings.

Reply:
@lolodomo - good point. I'll test this:
```java
        HttpClient httpClient = httpClientFactory.createHttpClient(LGWebOSBindingConstants.BINDING_ID);
        SSLContext sslContext = new SslContextFactory.Client(true).getSslContext();
        httpClient.getSslContextFactory().setSslContext(sslContext);
        this.webSocketClient = new WebSocketClient(httpClient);
```

Reply:
Result - named threads:

```
openhab> threads --list | grep lgwebos
104971 x pipe-grep lgwebos                                                                        x TIMED_WAITING x 180      x 160
104984 x OH-httpClient-lgwebos-104984                                                             x RUNNABLE      x 0        x 0
104985 x OH-httpClient-lgwebos-104985                                                             x RUNNABLE      x 5        x 0
104986 x OH-httpClient-lgwebos-104986                                                             x TIMED_WAITING x 1        x 0
104987 x OH-httpClient-lgwebos-104987                                                             x TIMED_WAITING x 0        x 0
104988 x OH-httpClient-lgwebos-104988                                                             x TIMED_WAITING x 0        x 0
```

But also:

```
javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.ssl.Alert.createSSLException(Alert.java:131) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:353) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:296) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:291) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:654) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:473) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:369) ~[?:?]
	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392) ~[?:?]
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1076) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1063) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1010) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.fill(SslConnection.java:654) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:168) ~[?:?]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:80) ~[?:?]
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:131) ~[?:?]
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:172) ~[?:?]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) ~[?:?]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:555) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:410) ~[?:?]
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:164) ~[?:?]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[?:?]
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) ~[?:?]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409) ~[bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) [bundleFile:9.4.46.v20220331]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) [bundleFile:9.4.46.v20220331]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:439) ~[?:?]
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306) ~[?:?]
	at sun.security.validator.Validator.validate(Validator.java:264) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:276) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:141) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:632) ~[?:?]
	... 28 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141) ~[?:?]
	at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126) ~[?:?]
	at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297) ~[?:?]
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:434) ~[?:?]
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306) ~[?:?]
	at sun.security.validator.Validator.validate(Validator.java:264) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:276) ~[?:?]
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:141) ~[?:?]
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:632) ~[?:?]
	... 28 more
```

Reply:
This might be the reason why I created a trust certificate manager. This is something also present in at least one another binding (because I remember I take my inspiration in another binding).

If you want to keep things simple, maybe keep your initial code (without httpClientFactory).

Let me know.

Reply:
I also see `OH-httpClient-lgwebos` threads. I am wondering what created these threads ?

Reply:
Normal, it is created by call to `webSocketFactory.createWebSocketClient`

Reply:
@jlaur : you can forget my proposal because `webSocketFactory.createWebSocketClient` is already creating the dedicated thread for the HTTP client.
I now understand that it makes no sense to use httpClientFactory if the aim to only to create a WEB socket client.
Sorry @jlaur for the lost time !

Reply:
Considering everything above, I would probably stick with the current implementation at the moment:
- It has been tested and verified to work.
- It seems simpler than having to override methods in `X509ExtendedTrustManager`.
- It's used by other bindings as well.

I still believe it may be worth revisiting, I'm just not sure right now how to create an implementation that's both simple and makes the best use of thread pools etc.

Reply:
Just to be precise, you replaced the call to `webSocketFactory.createWebSocketClient` by your new code, so the threads we are seeing are old threads (certainly from the previous version of the binding). A restart of OH would make them disappearing.

So no, these threads will no more be visible. HTTP client and WEB socket client will now be created by the binding without considering any OH core class.

Reply:
Confirmed after re-installing your version:
```
openhab> bundle:list -s | grep lgwebos
250 x Active   x  80 x 4.0.0.202302041217     x org.openhab.binding.lgwebos
openhab> threads --list | grep lgwebos
1038 x pipe-grep lgwebos                                                                         x RUNNABLE      x 4        x 0
openhab>
```